### PR TITLE
Add documentation for repositories and namespaces.

### LINKED ISSUE #2183: Document spack repo and repos.yaml
> FWIW, spack repo and repos.yaml still need to be documented a bit and needs a reference guide section. That can wait for another PR.

### PR DESCRIPTION
**Pull request by tgamblin**
Fixes #2183.

- [x] Document repositories
- [x] Document `repos.yaml` file
- [x] Document `spack repo` command
- [x] Document package namespaces
- [x] Simplify documentation TOC a bit by moving tutorial to basics section

@citibeth @adamjstewart @jrood-nrel @TomasPuverle

### CODE REVIEW

File: lib/spack/docs/repositories.rst [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+``package.py`` file and any patches or other files needed to build the
+package.
+
+
+---------------------
+``repos.yaml``
+---------------------
+
+Spack finds repositories by looking at the ``repos.yaml`` configuration
+file.  For more on the YAML format, and on how configuration file
```
Comment (citibeth) [inline lib/spack/docs/repositories.rst]: Looking at or looking for?  Oh, I see my confusion... `repos.yaml` is in the configuration area, but `repo.yaml` is at the top of every repo.  This is a "gotcha" that could probably be clarified by a carefully placed sentence in the docs.

File: lib/spack/docs/repositories.rst [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+.. code-block:: yaml
+
+  repos:
+  - /opt/local-repo
+  - $spack/var/spack/repos/builtin
+
+When you install a package with Spack, it searches these directories in
+order for an appropriate ``package.py`` file.  For example, if you type
+``spack install mpich``, Spack looks first in
```
Comment (citibeth) [inline lib/spack/docs/repositories.rst]: Is this precedence in the same order as elsewhere in Spack where there are multiple options listed?  (I believe it is; if not, we should be consistent everywhere).

File: lib/spack/docs/repositories.rst [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+``$spack/var/spack/repos/builtin/packages/mpich``, and uses the first
+valid package it finds.
+
+.. _note:
+
+  Currently, Spack only knows how to deal with repositories in the local
+  file system. Eventually we plan to support putting URLs in
+  ``repos.yaml`` so that you can easily host remote package repositories,
+  but that support is not implemented yet.
+
```
Comment (citibeth) [inline lib/spack/docs/repositories.rst]: Does "local filesystem" mean a HDD on the same machine, or can it be on an NFS share?

File: lib/spack/docs/repositories.rst [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+.. code-block:: console
+
+  $ cat var/spack/repos/builtin/repo.yaml
+  repo:
+    namespace: builtin
+
+The namespace doesn't **have** to correspond to the repository directory
+name; it's just convenient in this case.
+
```
Comment (citibeth) [inline lib/spack/docs/repositories.rst]: So to clarify... I can make my own repository using the `builtin` namespace?

There are now two systems of prioritization and overrides... one based on which repo a package is in, and one based on namespaces.  How do those two systems interract?  This is getting more complex than I first imagined...

File: lib/spack/docs/repositories.rst [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+---------------------
+Listing repositories
+---------------------
+
+Spack's :ref:`configuration system <configuration>` allows repository
+settings to come from ``repos.yaml`` files in many locations.  If you
+want to see the repositories registered as a result of all configuration
+files, use ``spack repo list``.
+
```
Comment (citibeth) [inline lib/spack/docs/repositories.rst]: Isn't there a general configuration command to merge configurations and show the result?  Why do we have a separate command here?

File: lib/spack/docs/repositories.rst (hunk 4 above)
Comment (citibeth) [inline lib/spack/docs/repositories.rst]: Is there any advice on upper vs. lower case repo and namespace names?  I would suggest recommending all lower-case...

File: lib/spack/docs/repositories.rst [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+.. _cmd-spack-repo-rm:
+
+^^^^^^^^^^^^^^^^^
+``spack repo rm``
+^^^^^^^^^^^^^^^^^
+
+You can remove an already-registered repository with ``spack repo rm``.
+This will work whether you pass the repository's namespace **or** its
+path.
+
```
Comment (citibeth) [inline lib/spack/docs/repositories.rst]: Does this (a) delete the repository directory, or does it (b) simply remove the repository from the appropriate `repos.py` file?  I think it should do just (b), because (a) could be quite dangerous.  Especially if the command does not ask for user confirmation. This is like how SmartGit removes Git repos: it removes them from its list of repos, it doesn't delete them.

When run, the command should also:
  a) Indicate which `repo.yaml` file was edited (allowing the user to undo if needed)
  b) Mention that the repo directory was not deleted; but print out an `rm -rf /repo/path` command so the user can cut-n-paste if she wants to delete it.

**Comment by citibeth** [comment]
Looks like some nice features.  As is so often the case, I manage to find some design issues when reading the documentation for the first time.

File: lib/spack/docs/repositories.rst (hunk 4 above)
Comment (TomasPuverle) [inline lib/spack/docs/repositories.rst]: I have a question about this, too.  As @citibeth mentioned, can multiple repos live in the same namespace, or does spack check that all repos have a unique namespace?  I feel the latter will become harder to enforce, especially as the possibility of "remote repos" (which you mentioned in the docs) exists.

In the documentation, the comment about "ordered" lists implies there is a preference while searching the repos for packages.  Could you clarify the following, please?
a) It is clearly legal (from the new docs) to have packages with the same name living in different repos.  When installing, is the "current" repo given precedence or is the search order always the same?  So if I install namespace.B  which needs A, will namespace.A be searched for first/given priority if, say, builtin.A exists?
b) Do the specs in package.py (such as depends_on) support the "namespace.A" syntax?  Is there a recommendation wrt this? (My initial feeling would be to, in general, avoid namespacing in package specs)

Comment (TomasPuverle) [inline lib/spack/docs/repositories.rst]: Actually, with respect to a) above: I have three repos, ordered as r1, r2, r3, all in the same namespace, all having B, A with B depending on A, is there a way to access the contents of r2?  
And if there were, and I install r2's B, which A will I get by default?

**Comment by TomasPuverle** [comment]
It's perhaps not entirely clear from the relative terseness of my comments/questions, but I think this is great!

**Comment by citibeth** [comment]
Thinking it through... I see the use case for multiple repos but not multiple namespaces.  Multiple repos already give an overriding capability if you want to override an existing package.  So does git fork.

File: lib/spack/docs/repositories.rst (hunk 2 above)
Comment (tgamblin) [inline lib/spack/docs/repositories.rst]: It is.

File: lib/spack/docs/repositories.rst (hunk 3 above)
Comment (tgamblin) [inline lib/spack/docs/repositories.rst]: I can clarify this -- either should work.

File: lib/spack/docs/repositories.rst (hunk 6 above)
Comment (tgamblin) [inline lib/spack/docs/repositories.rst]: (a) *would* be bad 😄.  It does (b).

**Comment by tgamblin** [comment]
> Thinking it through... I see the use case for multiple repos but not multiple namespaces. Multiple repos already give an overriding capability if you want to override an existing package. So does git fork.

@citibeth: It might not be clear from the documentation as written, but the point of the namespaces is to provide provenance, so that after a package is installed we know where it came from.  It's not perfect, in that people can choose similar namespaces, but it does let you differentiate between Spack's packages and your internal ones.

A repo is just a directory, which can be moved around the filesystem and shouldn't have to stay constant.  e.g., someone could move a repo from a local directory to a remote server, or from a local directory to a shared filesystem.  If we just had repos, and two packages had the same name, the hash would conflict. 

The *namespace* is an identifier that goes in the spec DAG to identify the package.  Without this, consider the following scenario:

1. User creates and adds a repo called, say, `mycompany`
1. User copies the built-in `package.py` for `mpich` into their repo, and modifies the build slightly
1. User tries to install `mpich`
1. The hash for this new `mpich` conflicts with what they've already installed.

The namespace avoids this -- the new `mpich` is actually `mycompany.mpich` and the builtin one is `builtin.mpich`.  They can coexist.

You could argue that when we start hashing `package.py` files, this will become unnecessary, but right now we don't hash `package.py` files (for reasons discussed in #2247).  I think even if we did that, you'd want a namespace so you don't have to wonder which repo a package with a particular hash came from.

Does that make sense?  Maybe I should put more of these details in the docs.

File: lib/spack/docs/repositories.rst (hunk 4 above)
Comment (tgamblin) [inline lib/spack/docs/repositories.rst]: > Is there any advice on upper vs. lower case repo and namespace names? I would suggest recommending all lower-case...

Not opposed to adding advice for lowercase.

**Comment by tgamblin** [comment]
@TomasPuverle 

> does spack check that all repos have a unique namespace? I feel the latter will become harder to enforce, especially as the possibility of "remote repos" (which you mentioned in the docs) exists.

It won't let you add a repo that has the same namespace as one already registered. It's a minimal protection, but at least it alerts you to the conflict.  If we do remote repos and the community is big enough, we could consider letting people "register" their repos.

> a) It is clearly legal (from the new docs) to have packages with the same name living in different repos. When installing, is the "current" repo given precedence or is the search order always the same? So if I install namespace.B which needs A, will namespace.A be searched for first/given priority if, say, builtin.A exists?

Every search is done from the top down, so, given some repos

```
r1: [bar]
r2: [foo]   # Package 'foo' depends_on('bar')
r3: [bar]
```

`spack install foo` will concretize to `r2.foo ^r1.bar`.  I think this is what most users will want -- if I override the `mpich` package in a repo, I *probably* meant to do it across the whole stack.  If I want to depend on a shadowed package, I can always do this:

```
spack install foo ^r3.bar
```

This is also consistent with the view of the repos the user sees via `spack list` -- `bar` only appears once in the list and shadowed packages don't get shown.  There should probably be a `-N` option for `spack list` like there is for `spack find`.

b) Do the specs in package.py (such as depends_on) support the "namespace.A" syntax? Is there a recommendation wrt this? (My initial feeling would be to, in general, avoid namespacing in package specs)

They do.  In fact, this is integrated with the parser, so you can refer to a spec by namespace anyplace that accepts a spec.  I would recommend *against* explicitly namespacing dependencies.  Maybe I should elaborate a bit.

I did not write this in the docs but you can also, within a package, do this:

```python
from spack.pkg.builtin.mpich import Mpich

class MyPackage(Mpich):
    ...
```

So the namespaces are actually Python namespaces that get tacked on under `spack.pkg`, and repo search actually uses the same mechanism as Python's path search.  The repository infrastructure is actually a custom python [importer](https://docs.python.org/3/reference/import.html), and I tried to write it so there weren't too many surprises about how the classes in `package.py` files actually get loaded.  

Early on, Spack just used `mpich.py` for the `mpich` package.  We decided that we needed packages to live in directories so package authors could add things like patch files, so we move to making them proper Python packages -- so `mpich` then lived in an `mpich` directory with an `__init__.py` in it.  People (users) really didn't like the `__init__.py` thing, and it made it hard to refer to the actual package *file* in conversation.  That an the desire to dynamically add/remove repos is where the current design came from.

It is probably obvious that going *too* far down this road is going to lead to having inter-repo dependencies (meta-problems!), and I *really* don't want to get into that mess.  The current design handles these goals:

1. Allow users to maintain their own packages separately from Spack.
1. Allow users to share these packages easily without the requirement to submit a package to the `builtin` repo (not done but this is what remote repos are for).
1. Allow users' packages to *depend on* `builtin` Spack packages.
1. Allow users to override `builtin` Spack packages if they want to.

I consider generalizing (3), i.e. supporting smooth installation of packages from repos with lots of dependencies on other repos, to be outside the current scope.  I think it could get really nasty and it would require re-engineering a bunch of stuff that works nicely *within* a repo right now.

TL;DR: don't use explicit namespaces in dependencies, at least not right now.

c) Going back to the "ordered list" comment, it's not clear in the docs if there is, in fact, a priority and in which order it is

Thanks!  I'll clarify this.

File: lib/spack/docs/repositories.rst (hunk 4 above)
Comment (tgamblin) [inline lib/spack/docs/repositories.rst]: @TomasPuverle: 

> Actually, with respect to a) above: I have three repos, ordered as r1, r2, r3, all in the same namespace, all having B, A with B depending on A, is there a way to access the contents of r2?
And if there were, and I install r2's B, which A will I get by default?

Not if they're all in the same namespace.  The namespace *is* the identifier for the repo.  Spack will yell at you, though, if you register two of these repos at once via `spack repo add`.  The hope is that people will choose good namespaces, which I think is all I can do.

I think this is clear from my prior comment, but If there were a way to do this, and you installed `r2`'s `B`, you'd get `r1`'s `A`.

**Comment by tgamblin** [comment]
@TomasPuverle:

> It's perhaps not entirely clear from the relative terseness of my comments/questions, but I think this is great!

Thanks!

**Comment by tgamblin** [comment]
@citibeth @TomasPuverle: If the comments above clarify everything for people, I'll update the docs.  let me know.

**Comment by TomasPuverle** [comment]
On 12/04/2016 02:47 PM, Todd Gamblin wrote:
> It won't let you add a repo that has the same namespace as one already 
> registered. It's a minimal protection, but at least it alerts you to 
> the conflict. If we do remote repos and the community is big enough, 
> we could consider letting people "register" their repos. 
Well, that addresses the majority of my concerns/questions.

>  If I want to depend on a shadowed package, I can always do this:
> |spack install foo ^r3.bar|
Perfect.

> I did not write this in the docs but you can also, within a package, 
> do this:
> from  spack.pkg.builtin.mpichimport  Mpich
>
> class  MyPackage(Mpich):
>      ...
>
> So the namespaces are actually Python namespaces that get tacked on 
> under |spack.pkg|, and repo search actually uses the same mechanism as 
> Python's path search. The repository infrastructure is actually a 
> custom python importer 
> <https://docs.python.org/3/reference/import.html>, and I tried to 
> write it so there weren't too many surprises about how the classes in 
> |package.py| files actually get loaded.
>
That is pretty cool.  One note thought (and I'm sure you know this, and 
realize that there are currently some Python 3 issues being worked out 
in spack) - the package import/finding infra changed quite significantly 
between Python 2 and 3.


> TL;DR: don't use explicit namespaces in dependencies, at least not 
> right now. 
Might it be worth mentioning this as a best practice in the docs? 
(Especially the packaging guide).

File: lib/spack/docs/repositories.rst (hunk 4 above)
Comment (adamjstewart) [inline lib/spack/docs/repositories.rst]: As far as upper-case vs. lower-case, I don't think it really matters. It's up to the user. No one else will ever even see it, so no one can complain. No reason to have to enforce this.

> The namespace doesn't **have** to correspond to the repository directory name

Is there any need for this feature? Is the `repo.yaml` only necessary to provide the name of the repo? Why not just use the directory name in all cases?

Comment (adamjstewart) [inline lib/spack/docs/repositories.rst]: Also, I think the answers to the questions @TomasPuverle raised about how the spec resolution mechanism handles namespaces should be added to the documentation.

File: lib/spack/docs/repositories.rst [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+---------------------
+Listing repositories
+---------------------
+
+Spack's :ref:`configuration system <configuration>` allows repository
+settings to come from ``repos.yaml`` files in many locations.  If you
+want to see the repositories registered as a result of all configuration
+files, use ``spack repo list``.
+
+.. _cmd-spack-repo-list:
```
Comment (adamjstewart) [inline lib/spack/docs/repositories.rst]: We established an unofficial convention that the [Command Index](http://spack.readthedocs.io/en/latest/command_index.html) should contain subcommands (`spack compiler`, `spack repo`) but not subsubcommands (`spack compiler add`, `spack compiler find`, `spack compiler remove`, ...) since they are always documented in the same area.

File: lib/spack/docs/repositories.rst [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+
+Once a repository is registered like this, you should be able to see its
+packages' names in the output of ``spack list``, and you should be able
+to build them using ``spack install <name>`` as you would with any
+builtin package.
+
+.. _cmd-spack-repo-rm:
+
+^^^^^^^^^^^^^^^^^
+``spack repo rm``
```
Comment (adamjstewart) [inline lib/spack/docs/repositories.rst]: Can you use `remove` instead of `rm`? I know it's an alias, but we have a section for `spack compiler remove` already. Just trying to be consistent.

File: lib/spack/docs/repositories.rst (hunk 7 above)
Comment (tgamblin) [inline lib/spack/docs/repositories.rst]: I think we need a new "conventions" section in the docs to list these things 😄

File: lib/spack/docs/repositories.rst (hunk 8 above)
Comment (tgamblin) [inline lib/spack/docs/repositories.rst]: Yeah I can do that.  I even thought about it.  Then I thought about what I would actually type and went with that, but `remove` is clearer.

**Comment by tgamblin** [comment]
Just wanted to say thanks to everyone for all the feedback.  I think the comments really help this PR.

File: lib/spack/docs/repositories.rst (hunk 7 above)
Comment (adamjstewart) [inline lib/spack/docs/repositories.rst]: We should also add a bullet point in `CONTRIBUTING.md` so that people know to add new conventions to the "conventions" section of the docs 😆

**Comment by TomasPuverle** [comment]
Hi Todd,

It occurred to me that your new namespace feature may actually provide (at least a partial) resolution to one of the "bigger" issues I've been reading about on the list and here: "Build vs Reuse".

First, I'm not sure if you've seen my suggestion about abstracting the "repository"; as a starting point, instead of "@system", there could actually be a repository (which, perhaps at least as a first pass, just reads packages.yaml). (Eventually, the idea would we interaction with "foreign" package managers but for now, this is irrelevant to this discussion.  However, I think my suggestion further illustrates the power of this abstraction).  

With this abstraction in place, foo@system - or perhaps better with the namespace support, system.openssl - would now become a package in its own right.

Let's introduce another repository.  This repository is called "installed", and reads ".spack.db", and has a list of packages which were installed by spack.

Now; some people prefer to build first, then search; others prefer to search first, then build.  By ordering the three "known" repositories - builtin, system and installed appropriately in repos.yaml, it may be possible to effectively select one behavior over another.

Furthermore, using namespaces, one would continue to be able select something different than default: spack install boost ^installed.zlib or spack install boost ^builtin.zlib.

Thanks,

Tom

**Comment by tgamblin** [comment]
Ok, I *think* this is updated per all the suggestions.  I added a lot of material and tried to clean up the organization.  I think it's good for now.

@adamjstewart: do you approve?

**Comment by tgamblin** [comment]
@TomasPuverle: in principle I like the idea of recycling the repo concept for externals and for different types of packages.

The issue I see with the way externals currently work is that externals really are stand-ins for particular `package.py` files.  When you build with an external `mpich`, Spack still uses the first `Package` class found to construct a `Spec` DAG, and there are places in the build where that matters.  In particular, see #1821 and #1875.  When client code refers to `spec['lapack'].libs`, we're really asking `var/spack/repos/builtin/packages/lapack/package.py` to tell us what the libraries are for LAPACK, and some packages actually have some built-in code to handle external LAPACK installs --- e.g., they  do a search for `liblapack.a` or `liblapack.so`.   The `builtin.lapack` namespace tells you which `Package` class was used for the build, even for externals.

So in that sense, these are orthogonal concepts, at least at the moment.  We could think about separating the concerns, e.g. by making a repo for externals and moving the searching code into it, but I am not sure it makes sense.  Users who make repositories right now can add any of their new packages as externals and add some (usually small) amount of code to handle any discovery that has to happen for other builds to use the package as an external.
